File: packages/testing/src/pact.ts
import type { BoundaryValue, JsonObject } from "@opensesame/os-domain";
/**
 * Property / Adversarial / Chaos / conTract helpers (see docs/validation/pact.md).
 */

/** Run `worker` concurrently and return how many times it reported success. */
export async function countConcurrentWins(
  worker: () => Promise<boolean> | boolean,
  n = 32,
): Promise<number> {
  const results = await Promise.all(
    Array.from({ length: n }, () => Promise.resolve().then(worker)),
  );
  return results.filter(Boolean).length;
}

/** Exclusive claim: only one concurrent caller may succeed. */
export async function assertExclusiveClaim(
  worker: () => Promise<boolean> | boolean,
  n = 32,
): Promise<void> {
  const wins = await countConcurrentWins(worker, n);
  if (wins !== 1) {
    throw new Error(`pact exclusive claim: expected 1 winner, got ${wins}`);
  }
}

/** Capacity fence: concurrent callers may succeed at most `max` times. */
export async function assertAtMostWins(
  worker: () => Promise<boolean> | boolean,
  max: number,
  n = 32,
): Promise<number> {
  const wins = await countConcurrentWins(worker, n);
  if (wins > max) {
    throw new Error(`pact capacity: expected at most ${max} wins, got ${wins}`);
  }
  return wins;
}

/**
 * Check-then-set mutant: two observers both see "absent" then both write.
 * Used to prove why the production path must not be a naked read/write.
 */
export function checkThenSetAdmitsDoubleClaim(): void {
  const keys = new Set<string>();
  const check = (key: string) => !keys.has(key);
  const set = (key: string) => {
    keys.add(key);
  };
  const c1 = check("d1");
  const c2 = check("d1");
  if (!(c1 && c2)) {
    throw new Error("pact mutant: expected both checks to pass");
  }
  set("d1");
  set("d1");
}

/** Production source must mention markers in order (mutation oracle). */
export function assertSourceOrder(src: string, ordered: string[]): void {
  const production =
    src.split("#[cfg(test)]")[0]?.split("\ndescribe(")[0] ?? src;
  let last = 0;
  for (const marker of ordered) {
    const pos = production.indexOf(marker, last);
    if (pos < 0) {
      throw new Error(`pact source oracle missing ${marker}`);
    }
    last = pos;
  }
}

const DEFAULT_SECRET_FIELDS = [
  "access_token",
  "refresh_token",
  "client_secret",
  "webhook_secret",
  "private_key",
  "claim_token",
  "claimToken",
] as const;

/** Contract: a JSON body must not carry secret-shaped keys (except `allow`). */
export function assertNoSecretFields(
  value: BoundaryValue,
  allow: readonly string[] = [],
): void {
  const allowed = new Set(allow);
  const blob = JSON.stringify(value);
  for (const leak of DEFAULT_SECRET_FIELDS) {
    if (allowed.has(leak)) continue;
    if (blob.includes(`"${leak}"`)) {
      throw new Error(`pact contract: secret field ${leak} present`);
    }
  }
}

/** Contract: documented HTTP statuses must include fail-closed codes. */
export function assertFailClosedStatuses(
  responses: JsonObject | undefined,
  required: string[] = ["401"],
): void {
  if (!responses) {
    throw new Error("pact contract: missing responses object");
  }
  for (const code of required) {
    if (!(code in responses)) {
      throw new Error(`pact contract: missing fail-closed status ${code}`);
    }
  }
}

/** Chaos: after a publish/partition failure, durable work must still be present. */
export async function assertDurableSurvivesPartition(
  durableCount: () => Promise<number> | number,
  publish: () => Promise<void>,
): Promise<void> {
  try {
    await publish();
  } catch {
    // partition is the scenario under test
  }
  const n = await durableCount();
  if (n < 1) {
    throw new Error("pact chaos: durable work dropped after partition");
  }
}

File: crates/host-core/src/pact.rs
/// Production source must mention `ordered` markers in that sequence.
///
/// # Panics
///
/// Panics when a marker is absent or appears out of order.
pub fn assert_source_order(src: &str, ordered: &[&str]) {
    let production = src.split("#[cfg(test)]").next().unwrap_or(src);
    let mut last = 0usize;
    for marker in ordered {
        let pos = production
            .get(last..)
            .and_then(|rest| rest.find(marker))
            .map_or_else(
                || panic!("pact source oracle missing {marker}"),
                |offset| last + offset,
            );
        last = pos;
    }
}

/// Exclusive insert: only one concurrent claimant wins.
///
/// # Panics
///
/// Panics if the test mutex is poisoned or the single-winner invariant fails.
pub fn exclusive_claim_is_single_winner() {
    use std::collections::HashSet;
    use std::sync::{Arc, Mutex};
    struct Kv {
        keys: Mutex<HashSet<String>>,
    }
    impl Kv {
        fn try_claim(&self, key: &str) -> bool {
            let mut g = self.keys.lock().expect("kv");
            g.insert(key.to_string())
        }
    }
    let kv = Arc::new(Kv {
        keys: Mutex::new(HashSet::new()),
    });
    let mut wins = 0usize;
    for _ in 0..64 {
        if kv.try_claim("d1") {
            wins += 1;
        }
    }
    assert_eq!(wins, 1, "try_claim must be exclusive");
}

/// Check-then-set is the mutant that exclusive claim exists to kill.
///
/// # Panics
///
/// Panics if the test mutex is poisoned or the modeled race is not observed.
pub fn check_then_set_admits_double_claim() {
    use std::collections::HashSet;
    use std::sync::Mutex;
    let keys = Mutex::new(HashSet::<String>::new());
    let check = |key: &str| !keys.lock().expect("kv").contains(key);
    let set = |key: &str| {
        keys.lock().expect("kv").insert(key.to_string());
    };
    let c1 = check("d1");
    let c2 = check("d1");
    assert!(c1 && c2, "mutant race: both checks pass");
    set("d1");
    set("d1");
    assert_eq!(keys.lock().expect("kv").len(), 1);
}
